### ISSUE TITLE
Bring in the docs repo

### ISSUE THREAD

**Issue by HarelM** [maplibre/maplibre-gl-js#2150]
## Motivation

The docs repo is using an npm package of this repo, which basically is saying that it is coupled with the code here.
In order to streamline the docs and also maybe allow using the examples as "debug pages" it is advised to move the docs into this repo.
This was somewhat discussed here: #2148

## Design Alternatives

Keep the current setup and continue this split which doesn't really make sense.

## Acceptance criteria for the bounty

1. A design of how everything is going to work is presented and accepted, in this design the following should be considered (but they don't have to be implemented) - if needed, a separate bounty can be set for this.
  1.1. How to support in the future the ability to see docs of previous versions
  1.2. Do we want to allow seeing the docs for the main version (not the released one), if so, how?
  1.3. How do these docs interact with the style-spec docs

### Design

- The docs will be using mkdocs material as the infrastructure for the docs, ~with the ability to support versioning using `mike`~
- The examples will be in the tests folder and copied (and markdown file will be added to then) when building the docs so that that they can be used both for the docs and for debug.
- The docs of the current version can be easily seen when developing locally, no need to publish them.
- All the API comments will use TSdocs that will be converted to markdown using typedoc.
- The style spec docs will be kept separate at this point and there will be a link to them like the current solution, we can decide if we want to use the same infrastructure for the style-spec in the future.

### Implementation tasks

1. Setup mkdocs material configuration
1. Docs CI is running on every PR to make sure the docs are building without errors or warnings.
1. The release CI pipeline is updated to publish the docs on every release.
1. All the jsdocs comments are converted to tsdoc comments
1. The maplibre.org documentation links are still working as before (to some extent)
1. The maplibre-gl-js-docs repo is archived

Link to bounty:
https://github.com/maplibre/maplibre/issues/191

**Comment by birkskyum** [comment]
All the JSDoc generation is based on the installed maplibre-gl npm package and parse through the typescript there. The only files we use from the maplibre-gl-js submodule are from within the style-spec folder, which would be moved out for this repo by #2149. I'm not trying to make a case for or against this task, but pointing to that it arguably invalidates what looks like the main motivation for this exercise:

>The docs repo is using a submodule of this repo, which basically is saying that it is coupled with the code here.

**Comment by HarelM** [comment]
No argue, but in general I don't see a benefit of keeping the docs outside when the docs are aimed only for this repo.
If the docs were serving native as well that might have been a different story, but they don't, so having them close would benefit us all I believe - for example when releasing a version everything will be aligned and won't require extra work, we can think if we want to ship the docs with the npm package for example, etc...

**Comment by birkskyum** [comment]
I agree that it might as well be together.

I hope it will allow us to know at the time of a PR if a change to maplibre-gl will break the docs. An example is the recent rollup cleanup in maplibre-gl-js, as it broke the submodule workaround for the style-spec in the docs (#1986) - and unless we manually investigate we wouldn't know untill an official release breaks the ci-pipeline on the main branch when deploying the docs.

**Comment by HarelM** [comment]
Good point, I'll add it to the accepted criteria.

**Comment by krishnaglodha** [comment]
Hi @HarelM , please let me know if the task is available, I'd like to contribute to it

**Comment by birkskyum** [comment]
For the style-spec, the docs are published from the last commit on main. I don't think it's a good idea here, because it would publish pre-releases, like 3.0.0-pre.4.

Assuming we go with a manual trigger, pre-release docs can still be previewed by running the docs locally.

**Comment by HarelM** [comment]
I guess we should consider a mechanism to show different versions...?

**Comment by birkskyum** [comment]
In that case, building the docs for maplibre < 3.0.0-pre.4 is actually easiest to do while the repos are still separate and the version of maplibre-gl can be adjusted in package.json.

The tools we have if we merge is:
- the changelog
- the ability to run any version >= 3.0.0-pre.4 locally after merging, by building docs locally
- optionally the ability to publish different versions of the docs, but only for >= 3.0.0-pre.4

**Comment by HarelM** [comment]
I would consider placing folders in the gothub pages branch and manually copy build versions prior to 3, after that make the release github action publish the relevant version to the relevant folder (maybe specific version + override latest folder or something), I think this will be the easiest to achieve.
Bottom line, let's not over engineer this, it's basically a static site that we should be able to copy around...

**Comment by birkskyum** [comment]
This is still available for anyone interested.

**Comment by HarelM** [comment]
I've updated the bounty size and removed the optional elements in the acceptance criteria, moved most of them to the design part of this issue to make sure we take into account future plans for the docs.

**Comment by birkskyum** [comment]
Some thoughts regarding the versioning - Because links are hardcoded in a static build, we can't move a static page to another directory after it is built (i.e. from / to /version/2.4.0), and also we wouldn't be able to change the base url (i.e. from maplibre-gl-js-docs to maplibre-gl-js) without rebuilding. In order to keep the ability to rebuild a version, what we need is the .json output from documentation.js - so it could make sense to store this .json for each version next to the html files in the gh-pages branch.

**Comment by HarelM** [comment]
This assumes we will continue to use documentation.js.
From my point of view, since the documentation is a generated static site we can store whatever the output is (assuming it is complete) in a folder, previous versions don't have to be perfect, they just need to work, and worst case you search and replace urls in the generated related version inside the gh-pages branch (if this scenario is interesting, which I'm not certain it is).
The aim here is simplicity mainly, much like older branches of the code, the main focus is the latest version, not an outdated one.
The only version that needs to know about the location of the previous version is the last one, I'm OK with redirecting the user out of the main page for older versions.

**Comment by birkskyum** [comment]
Excellent points - so no rebuilds will be needed, at most a find/replace

**Comment by HarelM** [comment]
Here are my finding regarding some experiments I've done, in continue to the following discussion: #2148.
I've create a branch here called typedoc-1.0.
In this branch there's a way to create markdown by running `npm run generate-docs`.
This will create a folder with all the classes, not very helpful.
Typedoc generated site is OK, but it can't be extended, so it's not a great experience and a lot of the classes are not interesting/important, so there might be a need to do something there as well.
I've looked a bit into the following options:
1. Jekyll - I've experimented with it, it requires some configurations, I'm not sure the configuration are worth the effort. pros - it's the default that github uses.
2. Hugo - There will be a need to edit the markdown, I don't see this as a valid option to be honest.
3. Mkdocs - It might be the easiest config to use, so that's an option I think, I haven't played around with it enough.
4. Docusaurus - It's a full react website, which I feel is an overkill, with npm packages, and a lot of configuration, it has a typedoc plugin, which might make things simpler. This is what deck.gl and react-map are using, it's the same technology in term of leveraging web usage. bottom line it has a lot of pro and a lot of cons.

After the above is selected there will be a need to create a script that will take all the examples pages (they will be hosted here and available to be use as a stand alone pages in this repo) and convert them to something the above "site" can present, it should be a relatively simple script.

There will also be a need to convert all the jsdos comments to TSDocs.

Versioning is will be done with "static" folders in the github pages.

I think this almost covers the design part of this issue.
I'll need to experiment with mkdocs and docusaurus a bit to better know what should be the way forward.

**Comment by HarelM** [comment]
After seeing how many files docusaurus added in order to run a simple static website I decided it is not the right solution.
mkdocs material on the other hand is super simple, well document and doesn't require much besides a simple configuration file.
I managed to make it work with the API and I've experimented with a simple example page, it should be simple to make it all work.
It also allows versioning (using `mike`) and dark/light out of the box.
Bottom line, I believe this should be the tool of choice for this problem.
There will be a need for a script to handle things related to the examples and maybe the public API classes selection.

**Comment by birkskyum** [comment]
Interesting findings. Just a note on typedoc so it's not missed in the evaluation - it's html output is not the best as you found too, but it has a JSON mode as well, with plugins for many ssg's like:

Docusaurus: https://www.npmjs.com/package/docusaurus-plugin-typedoc 
Markdown: https://www.npmjs.com/package/typedoc-plugin-markdown

**Comment by HarelM** [comment]
Yes, the markdown is the key here.
It is what docusaurus uses, and the plugin is under the same company, but I think docusaurus is too complicated.
I tend to think mkdocs solution can host the style spec as well by conversion the style spec to markdown straight from the v8.json file, but that's a different story...
I'll see if I can send a link tomorrow to the experiment I'm doing so you can see how it looks.

**Comment by HarelM** [comment]
So the current status is as follows:
I have brought in all the examples and examples' images to the typedoc-1.0 branch.
I created a small script that turns the examples html files to markdown.
The examples can be used to debug code, which is a huge plus.
There are just a few added dependencies, which is also great.
In order to test how the docs look the following docker command can be used (no need to install anything basically besides docker):
`docker run --rm -it -p 8000:8000 -v ${PWD}:/docs squidfunk/mkdocs-material`

I've deployed the site to a test repo, just so it can be viewed:
https://harelm.github.io/maplibre-docs-test/examples/

The API documentation is not as I hoped it would be, it's a bit too verbose to my taste, I've created an intro page to the API to help the reader get to the "important" parts of it (The public API - only contains the Map and FullScreenControl):
https://harelm.github.io/maplibre-docs-test/API/

Still to be done:
- [ ] Convert all the jsdocs to tsdocs by creating PRs against the current typedoc branch.
- [ ] Add favicon and maplibre icons in the right places (mkdocs config, should be easy)
- [ ] Add this to the CI so that it is published when there is a new release
- [ ] Archive the old repo (and redirect the old calls to the new docs?)
- [ ] Bring in the image generation script and update it
- [ ] Update the relevant readme file

But before we continue, we need to decide if this is a good enough solution.
I know the current site looks a bit better, but I think the benefits of having a synced docs and typescript code is more important, it will also help when developing to get the right intelisense.
It's also fully automatic (besides the `@group` that needs to be specified in the typedoc.json for a new group and used in the relevant tsdoc of a class to be included in the intro to the API).

**Comment by birkskyum** [comment]
I think this approach is very sound. If the api output of typedoc isn't as expected (you mentioned it's too verbose), an alternative way to make markdown for mkdocs consumption is the [api-extractor](https://github.com/microsoft/rushstack/tree/main/apps/api-extractor) followed by [api-documenter](https://github.com/microsoft/rushstack/tree/main/apps/api-documenter) - i don't know if it's more or less verbose though.

I wasn't able to generate the docs locally, leaving me with this output.

<img width="510" alt="Screenshot 2023-06-26 at 17 55 01" src="https://github.com/maplibre/maplibre-gl-js/assets/74932975/fad726bc-17f2-424b-9bed-b553dfa5ef57">

**Comment by HarelM** [comment]
Did you run npm i and npm generate-docs?

**Comment by birkskyum** [comment]
Yes, also tried running. npm run build-dist in between just in case. It seems like the npm generate-docs print a large amount of errors to the terminal like this:

```shell
staging/maplibregl/worker.js:1:31431 - error TS2339: Property 'getLeaves' does not exist on type 'z | de'.
  Property 'getLeaves' does not exist on type 'de'.

1 define(["./shared"],(function(e){"use strict";class t{constructor(e){this.keyCache={},e&&this.replace(e)}replace(e){this._layerConfigs={},this._layers={},this.update(e,[])}update(t,o){for(const o of t){this._layerConfigs[o.id]=o;const t=this._layers[o.id]=e.createStyleLayer(o);t._featureFilter=e.createFilter(t.filter),this.keyCache[o.id]&&delete this.keyCache[o.id]}for(const e of o)delete this.keyCache[e],delete this._layerConfigs[e],delete this._layers[e];this.familiesBySource={};const i=e.groupByLayout(Object.values(this._layerConfigs),this.keyCache);for(const e of i){const t=e.map((e=>this._layers[e.id])),o=t[0];if("none"===o.visibility)continue;const i=o.source||"";let r=this.familiesBySource[i];r||(r=this.familiesBySource[i]={});const s=o.sourceLayer||"_geojsonTileLayer";let n=r[s];n||(n=r[s]=[]),n.push(t)}}}class o{constructor(t){const o={},i=[];for(const e in t){const r=t[e],s=o[e]={};for(const e in r){const t=r[+e];if(!t||0===t.bitmap.width||0===t.bitmap.height)continue;const o={x:0,y:0,w:t.bitmap.width+2,h:t.bitmap.height+2};i.push(o),s[e]={rect:o,metrics:t.metrics}}}const{w:r,h:s}=e.potpack(i),n=new e.AlphaImage({width:r||1,height:s||1});for(const i in t){const r=t[i];for(const t in r){const s=r[+t];if(!s||0===s.bitmap.width||0===s.bitmap.height)continue;const a=o[i][t].rect;e.AlphaImage.copy(s.bitmap,n,{x:0,y:0},{x:a.x+1,y:a.y+1},s.bitmap)}}this.image=n,this.positions=o}}e.register("GlyphAtlas",o);class i{constructor(t){this.tileID=new e.OverscaledTileID(t.tileID.overscaledZ,t.tileID.wrap,t.tileID.canonical.z,t.tileID.canonical.x,t.tileID.canonical.y),this.uid=t.uid,this.zoom=t.zoom,this.pixelRatio=t.pixelRatio,this.tileSize=t.tileSize,this.source=t.source,this.overscaling=this.tileID.overscaleFactor(),this.showCollisionBoxes=t.showCollisionBoxes,this.collectResourceTiming=!!t.collectResourceTiming,this.returnDependencies=!!t.returnDependencies,this.promoteId=t.promoteId}parse(t,i,s,n,a){this.status="parsing",this.data=t,this.collisionBoxArray=new e.CollisionBoxArray;const l=new e.DictionaryCoder(Object.keys(t.layers).sort()),c=new e.FeatureIndex(this.tileID,this.promoteId);c.bucketLayerIDs=[];const u={},h={featureIndex:c,iconDependencies:{},patternDependencies:{},glyphDependencies:{},availableImages:s},d=i.familiesBySource[this.source];for(const o in d){const i=t.layers[o];if(!i)continue;1===i.version&&e.warnOnce(`Vector tile source "${this.source}" layer "${o}" does not use vector tile spec v2 and therefore may have some rendering errors.`);const n=l.encode(o),a=[];for(let e=0;e<i.length;e++){const t=i.feature(e),r=c.getId(t,o);a.push({feature:t,id:r,index:e,sourceLayerIndex:n})}for(const t of d[o]){const o=t[0];o.source!==this.source&&e.warnOnce(`layer.source = ${o.source} does not equal this.source = ${this.source}`),o.minzoom&&this.zoom<Math.floor(o.minzoom)||o.maxzoom&&this.zoom>=o.maxzoom||"none"!==o.visibility&&(r(t,this.zoom,s),(u[o.id]=o.createBucket({index:c.bucketLayerIDs.length,layers:t,zoom:this.zoom,pixelRatio:this.pixelRatio,overscaling:this.overscaling,collisionBoxArray:this.collisionBoxArray,sourceLayerIndex:n,sourceID:this.source})).populate(a,h,this.tileID.canonical),c.bucketLayerIDs.push(t.map((e=>e.id))))}}let p,f,g,m;const y=e.mapObject(h.glyphDependencies,(e=>Object.keys(e).map(Number)));Object.keys(y).length?n.send("getGlyphs",{uid:this.uid,stacks:y,source:this.source,tileID:this.tileID,type:"glyphs"},((e,t)=>{p||(p=e,f=t,w.call(this))})):f={};const v=Object.keys(h.iconDependencies);v.length?n.send("getImages",{icons:v,source:this.source,tileID:this.tileID,type:"icons"},((e,t)=>{p||(p=e,g=t,w.call(this))})):g={};const x=Object.keys(h.patternDependencies);function w(){if(p)return a(p);if(f&&g&&m){const t=new o(f),i=new e.ImageAtlas(g,m);for(const o in u){const n=u[o];n instanceof e.SymbolBucket?(r(n.layers,this.zoom,s),e.performSymbolLayout({bucket:n,glyphMap:f,glyphPositions:t.positions,imageMap:g,imagePositions:i.iconPositions,showCollisionBoxes:this.showCollisionBoxes,canonical:this.tileID.canonical})):n.hasPattern&&(n instanceof e.LineBucket||n instanceof e.FillBucket||n instanceof e.FillExtrusionBucket)&&(r(n.layers,this.zoom,s),n.addFeatures(h,this.tileID.canonical,i.patternPositions))}this.status="done",a(null,{buckets:Object.values(u).filter((e=>!e.isEmpty())),featureIndex:c,collisionBoxArray:this.collisionBoxArray,glyphAtlasImage:t.image,imageAtlas:i,glyphMap:this.returnDependencies?f:null,iconMap:this.returnDependencies?g:null,glyphPositions:this.returnDependencies?t.positions:null})}}x.length?n.send("getImages",{icons:x,source:this.source,tileID:this.tileID,type:"patterns"},((e,t)=>{p||(p=e,m=t,w.call(this))})):m={},w.call(this)}}function r(t,o,i){const r=new e.EvaluationParameters(o);for(const e of t)e.recalculate(r,i)}function s(t,o){const i=e.getArrayBuffer(t.request,((t,i,r,s)=>{t?o(t):i&&o(null,{vectorTile:new e.vectorTile.VectorTile(new e.Protobuf(i)),rawData:i,cacheControl:r,expires:s})}));return()=>{i.cancel(),o()}}class n{constructor(e,t,o,i){this.actor=e,this.layerIndex=t,this.availableImages=o,this.loadVectorData=i||s,this.loading={},this.loaded={}}loadTile(t,o){const r=t.uid;this.loading||(this.loading={});const s=!!(t&&t.request&&t.request.collectResourceTiming)&&new e.RequestPerformance(t.request),n=this.loading[r]=new i(t);n.abort=this.loadVectorData(t,((t,i)=>{if(delete this.loading[r],t||!i)return n.status="done",this.loaded[r]=n,o(t);const a=i.rawData,l={};i.expires&&(l.expires=i.expires),i.cacheControl&&(l.cacheControl=i.cacheControl);const c={};if(s){const e=s.finish();e&&(c.resourceTiming=JSON.parse(JSON.stringify(e)))}n.vectorTile=i.vectorTile,n.parse(i.vectorTile,this.layerIndex,this.availableImages,this.actor,((t,i)=>{if(t||!i)return o(t);o(null,e.extend({rawTileData:a.slice(0)},i,l,c))})),this.loaded=this.loaded||{},this.loaded[r]=n}))}reloadTile(e,t){const o=this.loaded,i=e.uid,r=this;if(o&&o[i]){const s=o[i];s.showCollisionBoxes=e.showCollisionBoxes;const n=(e,o)=>{const i=s.reloadCallback;i&&(delete s.reloadCallback,s.parse(s.vectorTile,r.layerIndex,this.availableImages,r.actor,i)),t(e,o)};"parsing"===s.status?s.reloadCallback=n:"done"===s.status&&(s.vectorTile?s.parse(s.vectorTile,this.layerIndex,this.availableImages,this.actor,n):n())}}abortTile(e,t){const o=this.loading,i=e.uid;o&&o[i]&&o[i].abort&&(o[i].abort(),delete o[i]),t()}removeTile(e,t){const o=this.loaded,i=e.uid;o&&o[i]&&delete o[i],t()}}class a{constructor(){this.loaded={}}loadTile(t,o){const{uid:i,encoding:r,rawImageData:s}=t,n=e.isImageBitmap(s)?this.getImageData(s):s,a=new e.DEMData(i,n,r);this.loaded=this.loaded||{},this.loaded[i]=a,o(null,a)}getImageData(t){this.offscreenCanvas&&this.offscreenCanvasContext||(this.offscreenCanvas=new OffscreenCanvas(t.width,t.height),this.offscreenCanvasContext=this.offscreenCanvas.getContext("2d",{willReadFrequently:!0})),this.offscreenCanvas.width=t.width,this.offscreenCanvas.height=t.height,this.offscreenCanvasContext.drawImage(t,0,0,t.width,t.height);const o=this.offscreenCanvasContext.getImageData(-1,-1,t.width+2,t.height+2);return this.offscreenCanvasContext.clearRect(0,0,this.offscreenCanvas.width,this.offscreenCanvas.height),new e.RGBAImage({width:o.width,height:o.height},o.data)}removeTile(e){const t=this.loaded,o=e.uid;t&&t[o]&&delete t[o]}}function l(e,t){if(0!==e.length){c(e[0],t);for(var o=1;o<e.length;o++)c(e[o],!t)}}function c(e,t){for(var o=0,i=0,r=0,s=e.length,n=s-1;r<s;n=r++){var a=(e[r][0]-e[n][0])*(e[n][1]+e[r][1]),l=o+a;i+=Math.abs(o)>=Math.abs(a)?o-l+a:a-l+o,o=l}o+i>=0!=!!t&&e.reverse()}var u=e.getDefaultExportFromCjs((function e(t,o){var i,r=t&&t.type;if("FeatureCollection"===r)for(i=0;i<t.features.length;i++)e(t.features[i],o);else if("GeometryCollection"===r)for(i=0;i<t.geometries.length;i++)e(t.geometries[i],o);else if("Feature"===r)e(t.geometry,o);else if("Polygon"===r)l(t.coordinates,o);else if("MultiPolygon"===r)for(i=0;i<t.coordinates.length;i++)l(t.coordinates[i],o);return t}));const h=e.vectorTile.VectorTileFeature.prototype.toGeoJSON;var d={exports:{}},p=e.pointGeometry,f=e.vectorTile.VectorTileFeature,g=m;function m(e,t){this.options=t||{},this.features=e,this.length=e.length}function y(e,t){this.id="number"==typeof e.id?e.id:void 0,this.type=e.type,this.rawGeometry=1===e.type?[e.geometry]:e.geometry,this.properties=e.tags,this.extent=t||4096}m.prototype.feature=function(e){return new y(this.features[e],this.options.extent)},y.prototype.loadGeometry=function(){var e=this.rawGeometry;this.geometry=[];for(var t=0;t<e.length;t++){for(var o=e[t],i=[],r=0;r<o.length;r++)i.push(new p(o[r][0],o[r][1]));this.geometry.push(i)}return this.geometry},y.prototype.bbox=function(){this.geometry||this.loadGeometry();for(var e=this.geometry,t=1/0,o=-1/0,i=1/0,r=-1/0,s=0;s<e.length;s++)for(var n=e[s],a=0;a<n.length;a++){var l=n[a];t=Math.min(t,l.x),o=Math.max(o,l.x),i=Math.min(i,l.y),r=Math.max(r,l.y)}return[t,i,o,r]},y.prototype.toGeoJSON=f.prototype.toGeoJSON;var v=e.pbf,x=g;function w(e){var t=new v;return function(e,t){for(var o in e.layers)t.writeMessage(3,S,e.layers[o])}(e,t),t.finish()}function S(e,t){var o;t.writeVarintField(15,e.version||1),t.writeStringField(1,e.name||""),t.writeVarintField(5,e.extent||4096);var i={keys:[],values:[],keycache:{},valuecache:{}};for(o=0;o<e.length;o++)i.feature=e.feature(o),t.writeMessage(2,b,i);var r=i.keys;for(o=0;o<r.length;o++)t.writeStringField(3,r[o]);var s=i.values;for(o=0;o<s.length;o++)t.writeMessage(4,T,s[o])}function b(e,t){var o=e.feature;void 0!==o.id&&t.writeVarintField(1,o.id),t.writeMessage(2,I,e),t.writeVarintField(3,o.type),t.writeMessage(4,k,o)}function I(e,t){var o=e.feature,i=e.keys,r=e.values,s=e.keycache,n=e.valuecache;for(var a in o.properties){var l=o.properties[a],c=s[a];if(null!==l){void 0===c&&(i.push(a),s[a]=c=i.length-1),t.writeVarint(c);var u=typeof l;"string"!==u&&"boolean"!==u&&"number"!==u&&(l=JSON.stringify(l));var h=u+":"+l,d=n[h];void 0===d&&(r.push(l),n[h]=d=r.length-1),t.writeVarint(d)}}}function M(e,t){return(t<<3)+(7&e)}function P(e){return e<<1^e>>31}function k(e,t){for(var o=e.loadGeometry(),i=e.type,r=0,s=0,n=o.length,a=0;a<n;a++){var l=o[a],c=1;1===i&&(c=l.length),t.writeVarint(M(1,c));for(var u=3===i?l.length-1:l.length,h=0;h<u;h++){1===h&&1!==i&&t.writeVarint(M(2,u-1));var d=l[h].x-r,p=l[h].y-s;t.writeVarint(P(d)),t.writeVarint(P(p)),r+=d,s+=p}3===i&&t.writeVarint(M(7,1))}}function T(e,t){var o=typeof e;"string"===o?t.writeStringField(1,e):"boolean"===o?t.writeBooleanField(7,e):"number"===o&&(e%1!=0?t.writeDoubleField(3,e):e<0?t.writeSVarintField(6,e):t.writeVarintField(5,e))}d.exports=w,d.exports.fromVectorTileJs=w,d.exports.fromGeojsonVt=function(e,t){t=t||{};var o={};for(var i in e)o[i]=new x(e[i].features,t),o[i].name=i,o[i].version=t.version,o[i].extent=t.extent;return w({layers:o})},d.exports.GeoJSONWrapper=x;var C=e.getDefaultExportFromCjs(d.exports);const _={minZoom:0,maxZoom:16,minPoints:2,radius:40,extent:512,nodeSize:64,log:!1,generateId:!1,reduce:null,map:e=>e},D=Math.fround||(O=new Float32Array(1),e=>(O[0]=+e,O[0]));var O;const L=3,F=5,E=6;class z{constructor(e){this.options=Object.assign(Object.create(_),e),this.trees=new Array(this.options.maxZoom+1),this.stride=this.options.reduce?7:6,this.clusterProps=[]}load(e){const{log:t,minZoom:o,maxZoom:i}=this.options;t&&console.time("total time");const r=`prepare ${e.length} points`;t&&console.time(r),this.points=e;const s=[];for(let t=0;t<e.length;t++){const o=e[t];if(!o.geometry)continue;const[i,r]=o.geometry.coordinates,n=D(B(i)),a=D(Z(r));s.push(n,a,1/0,t,-1,1),this.options.reduce&&s.push(0)}let n=this.trees[i+1]=this._createTree(s);t&&console.timeEnd(r);for(let e=i;e>=o;e--){const o=+Date.now();n=this.trees[e]=this._createTree(this._cluster(n,e)),t&&console.log("z%d: %d clusters in %dms",e,n.numItems,+Date.now()-o)}return t&&console.timeEnd(
```

**Comment by HarelM** [comment]
These errors are from the staging folder, delete it first.
I'll try code spaces to see on a clean env.

**Comment by HarelM** [comment]
I've tried that on a clean environment using codespaces and it works as expected.
Besides the fact that when the examples folder doesn't exist the script fails, I'll push a commit to fix that.
I've also opened a PR for fixing the staging typescript issue: #2731.

**Comment by HarelM** [comment]
I've pushed a commit to fix the missing examples folder issue and the staging problem, should be ready to work now.

**Comment by HarelM** [comment]
I've opened a PR with things that won't affect this repo in anyway until we finish with the jsdoc - tsdoc migration which will happen on a different branch with small PRs against it for easy review.
After this migration there will be a need to configure the CI.
I've moved the instructions of the docs to the readme under documentation and diagrams folder.

**Comment by HarelM** [comment]
The PR was merged.
Remaining tasks are described above in the checkboxes.
I've opened a branch to allow collaboration around jsdocs tsdocs migration: called `jsdocs-tsdocs`.

**Comment by krishnaglodha** [comment]
I'd like to take up this task

**Comment by HarelM** [comment]
Great @krishnaglodha. I have assigned the issue to you.
I have also updated the acceptance criteria with all the information needed.
Please make sure to open PRs again the jsdocs-tsdocs branch.
Once all the JSdocs are migrated, there are no build warnings and docs build are part of the PR CI we will merge and focus on completing the release CI to publish the docs (making sure it stores the previous version).

If any of this is not clear or well defined please let me know.

**Comment by HarelM** [comment]
Just for reference the date of publishing version 1.15.2 is 2021-12-07 and 2.4 is 2022-08-27.
I thought about trying to get the docs from the gh pages branch of the docs repo, but I think the docs are too hard coded to run in their current location, so I don't think there will be an easy way show these old versions' docs in the new docs.
I also have some second thoughts about adding versioning to the docs as I generally think we should promote using the latest version.
IDK... we still have a few days to decide...
@birkskyum, @krishnaglodha any thoughts?

**Comment by birkskyum** [comment]
I think we can be okay with the latest, due to the very limited changes to the public api. I find a thorough changelog and migration info to be more useful than full docs of and old repo version.

**Comment by HarelM** [comment]
Good point, we can easily copy the changelog into the docs upon build (or maintain it there) to include that in the docs' site as well.

**Comment by HarelM** [comment]
I have removed versioning from the acceptance criteria and design proposal.

**Comment by HarelM** [comment]
I've split the CI part out of this issue into #2776.
This leaves this issue with only TSDocs-JSDocs migration.
I can't split the bounty well without creating a new issue that I'm not sure how to define.
This issue should be with L + M bounty (750$).

**Comment by HarelM** [comment]
Most of the work was completed, so I'm removing the bounties.
